app/routes.py

```diff
@@ -1 +1,272 @@
+        completed_at=completed_at
+        if request_body["completed_at"] is not None else None
+    )
+
+    db.session.add(new_task)
+    db.session.commit()
+
+    rsp = {"task": new_task.get_dict()}
+
+    return jsonify(rsp), 201
+
+
+@tasks_bp.route("", methods=["GET"])
+def get_all_tasks():
+    sort_query = request.args.get("sort")
+    tasks = Task.query.all()
+
+    if sort_query == "asc":
+        tasks = Task.query.order_by(sqlalchemy.asc(Task.title))
+    elif sort_query == "desc":
+        tasks = Task.query.order_by(sqlalchemy.desc(Task.title))
+
+    tasks_response = []
+    for task in tasks:
+        tasks_response.append(task.get_dict())
+
+    return jsonify(tasks_response), 200
+
+
+@tasks_bp.route("/<task_id>", methods=["GET"])
+def get_one_task(task_id):
+    task = validate_task(task_id)
+    rsp = {"task" : task.get_dict()}
+
+    return jsonify(rsp), 200
+
+
+@tasks_bp.route("/<task_id>", methods=["PUT"])
+def update_one_task(task_id):
+    task = validate_task(task_id)
+
+    if not request.is_json:
+        return {"msg" : "Missing JSON request body"}, 400
+
+    request_body = request.get_json()
+    try:
+        task.title = request_body["title"]
+        task.description = request_body["description"]
+    except KeyError:
+        return {
+            "msg": "Update failed due to missing data. Title, Description are required!"
+        }, 400
+
+    db.session.commit()
+
+    rsp = {"task" : task.get_dict()}
+    return jsonify(rsp), 200
+
+
+@tasks_bp.route("/<task_id>", methods=["DELETE"])
+def delete_one_task(task_id):
+    task = validate_task(task_id)
+
+    db.session.delete(task)
+    db.session.commit()
+
+    rsp = {"details": f'Task {task_id} "{task.title}" successfully deleted'}
+    return jsonify(rsp), 200
+
+@tasks_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_task_complete(task_id):
+    task = validate_task(task_id)
+    task.completed_at = datetime.now()
+    db.session.commit()
+    
+    Slack_Url = 'https://slack.com/api/chat.postMessage'
+    message = {"text": f"Someone just completed the task {task.title}", "channel": "C049FQLJTBN"}
+    token = os.environ.get('SLACKBOT-TOKEN')
+    requests.post(Slack_Url, json=message, headers={"Authorization": f"Bearer {token}"})
+
+    rsp = {"task": task.get_dict()}
+    return jsonify(rsp), 200
+
+
+@tasks_bp.route("/<task_id>/mark_incomplete", methods=["PATCH"])
+def mark_task_incomplete(task_id):
+    task = validate_task(task_id)
+    task.completed_at = None
+
+    db.session.commit()
+
+    rsp = {"task": task.get_dict()}
+    return jsonify(rsp), 200
+
+
+def validate_task(task_id):
+    try:
+        task_id = int(task_id)
+    except:
+        rsp = {"msg": f"Task with id #{task_id} is invalid."}
+        abort(make_response(rsp, 400))
+    
+    task = Task.query.get(task_id)
+
+    if not task:
+        rsp = {"msg": f"Task with id #{task_id} is not found!"}
+        abort(make_response(rsp, 404))
+    return task
```

Comment:
Great helper functions. Typically helper methods are placed in a separate file like routes_helpers. This just helps keep our model clean and separates our concerns